### ISSUE TITLE
Add ability to filter (skip) download of specific fragments

### ISSUE THREAD

**Issue by ischmidt20** [yt-dlp/yt-dlp#4211]
### Checklist

- [X] I'm requesting a feature unrelated to a specific site
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've verified that I'm running yt-dlp version **2022.06.22.1** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Description

Couldn't find any information anywhere else so asking here. I think it would be a very beneficial feature to allow for the skipping of fragments in the hls/dash downloader, either by index or by matching the URL to a regex (or both). This would come in very handy, especially for videos with many `#EXT-X-DISCONTINUITY`s.  For example, I have a playlist like:

```
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:6
#EXT-X-PLAYLIST-TYPE:EVENT
#EXT-X-MEDIA-SEQUENCE:0
#EXT-X-DISCONTINUITY
#EXT-X-DISCONTINUITY
#EXT-X-DISCONTINUITY
#EXT-X-DISCONTINUITY
#EXT-X-DISCONTINUITY
#EXT-X-DISCONTINUITY
#EXT-X-DISCONTINUITY
#EXT-X-PROGRAM-DATE-TIME:2022-06-26T19:59:59.253Z
#EXT-X-KEY:METHOD=NONE
#EXTINF:5.005,
https://dai.google.com/linear/pods/v1/seg/cmsid/2473515/video/7237a134-b40b-4b95-b5f8-53138971fe48/pod/4/profile/5600K/36.ts?pd=210010&sd=5005&so=175175&cust_params=evnt%3D0%26podID%3D4
#EXTINF:1.568233,
https://dai.google.com/linear/pods/v1/seg/cmsid/2473515/video/7237a134-b40b-4b95-b5f8-53138971fe48/pod/4/profile/5600K/37.ts?pd=210010&sd=1568&so=180180&cust_params=evnt%3D0%26podID%3D4&last=true
#EXT-X-DISCONTINUITY
#EXT-X-KEY:METHOD=AES-128,URI="https://playback.svcs.mlb.com/silk/events/98dc33c2-1c33-46f5-9571-c23fbf6c2c95/media/b2babd8e-5de3-42ef-b139-e93075891cc9/keys/c25fe137-47c5-422e-a59f-fea8820f7e5d",IV=0xB076CAF2AB75251CAD08EB89954DB438
#EXTINF:5.005,
177/20/00/05_872.ts
#EXTINF:5.005,
177/20/00/10_877.ts
#EXTINF:5.005,
177/20/00/15_882.ts
...
```

The segments with `https://dai.google.com` are advertisements, and not only would it be great to have these simply skipped, but in this case, the default postprocessor arguments result in a video that is only the advertisements and nothing else, due to the way ffmpeg handles the different streams. Allowing users to filter out fragments meeting certain criteria would greatly assist in the ease of downloading funky streams.

e.g. `--skip-fragments:regex 'dai.google.com'` or `--skip-fragments:index 1:100`

Considering the hls downloader already automatically filters out uplynk ads and the like, hopefully this wouldn't be too difficult to implement. Happy to help with this too if there is interest.

**Comment by pukkandan** [comment]
Do you have an example URL? yt-dlp has other mechanisms to skip/split ads. I want to check why they are insufficient for the usecase

**Comment by ischmidt20** [comment]
Basically any game at MLB.tv, might not be geolocked but unfortunately it is a subscription service and I cannot share my login. But here is an example in any case: https://www.mlb.com/tv/g663387

I have attached a full playlist here. To actually download it, you would need an auth header for the decryption key, but hopefully this gives you some idea about the structure. The pointer to the playlist is https://hlslive-akc-ewr1.media.mlb.com/hdnts=exp=1656470679~acl=/*~id=00u734ah0v7ecfdsl356~data=b2babd8e-5de3-42ef-b139-e93075891cc9~hmac=58898b4599917ec7148cee693e881496636cd6039437dc3abef681325bee84bc/b719012ac1b6e1e377d803b2a8fae863/va01/mlb/2022/06/26/Home_VIDEO_spa_Seattle_Mariners_Los_Ange_20220626_1656266855032/5600K/5600_complete_gdfp2-trimmed.m3u8
[5600_complete_gdfp2-trimmed.m3u.txt](https://github.com/yt-dlp/yt-dlp/files/8997345/5600_complete_gdfp2-trimmed.m3u.txt)

**Comment by ischmidt20** [comment]
I can easily rig my local version of yt-dlp by modifying the first of these 3 lines to be `if ad_frag_next or 'dai.google.com' in line:`

https://github.com/yt-dlp/yt-dlp/blob/962ffcf89c8d935410391fbea3580688aafe76d7/yt_dlp/downloader/hls.py#L127-L129

But obviously that is not behavior that should be merged into the actual software, although that condition could easily be determined by user input

**Comment by pukkandan** [comment]
Since DAIs are becoming more popular, we'll need some way to deal with them anyway (See https://github.com/yt-dlp/yt-dlp/issues/1089) - but I'm not sure how.

While the solution proposed in OP can help advanced users, it is useless for most. It is also unsuitable for automation. That said, I am not opposed to adding this feature anyway. A new option is probably unnecessary. We could implement this into `--download-sections`. Maybe something like: `--download-sections "$100-500,700-inf"`

A workaround for now is to use `--hls-split-discontinuity` and then manually concat the formats

**Comment by ischmidt20** [comment]
Agreed on all counts. Could be helpful for any extractors that need to pass those options along (for example, I have been thinking about writing one for the aforementioned MLB.tv, but I haven't found the time for it yet).

And thanks for introducing me to those new arguments, I suspect `--hls-split-discontinuity` will be very useful!
